### PR TITLE
fix nimscript handling when the target file does not exist

### LINKED ISSUE #17600: `nim sometask main.nonexistant` works even if main.nonexistant doesn't exist
### Example

main.nims:
```nim
task foo, "some foo":
  echo "in task foo"
```

nim foo main.nims # case 1
nim foo main.nonexistant # case2
nim foo main # case3

ditto with things like `help`

### Current Output
these all work and run the foo task

### Expected Output
only case1 should work
case2 is a clear bug, since main.nonexistant doesn't exist
case3 is IMO a bug, IMO this shouldn't be legal, it's too loose (given that task foo is unconstrained and, after https://github.com/nim-lang/Nim/pull/17596, the extension is unconstrained too, including no extension)

### Additional Information
* 1.5.1 8ee0eda84156c8d9074b7a42cdedc209cffdc4a4

### PR DESCRIPTION
**Pull request by sivchari**
This PR fix #17600

main.nims:
```nim
task foo, "some foo":
  echo "in task foo"
```

case1
```sh
nim foo main.nims
```

output
```sh
in task foo
```

case2
```sh
nim foo main.nonexistant
```

output
```sh
this file is not existed
Error: invalid command: foo
```

case3
```sh
nim foo main
```

output
```sh
this file is not existed
Error: invalid command: foo
```

this approach decides if nimscript file(*.nims) exists.
so case1 passes, but case2 and case3 fail for nimscript file does not exist.
case3 is IMO a bug, too.  so that I did above approach.
however, this is contradict to #17596.  so, I think that we might have to talk about nimscript specification before this PR.

### CODE REVIEW

**Comment by sivchari** [comment]
I made a force push because I made a mistake with the account to push.
Sorry about that.

File: compiler/nimconf.nim [hunk 1]
```diff
@@ -254,6 +254,10 @@ proc loadConfigs*(cfg: RelativeFile; cache: IdentCache; conf: ConfigRef; idgen:
     if isMain and optWasNimscript in conf.globalOptions:
       if conf.projectIsStdin: s = stdin.llStreamOpen
       elif conf.projectIsCmd: s = llStreamOpen(conf.cmdInput)
+    if conf.projectFull.string != "" and fileExists(conf.projectFull) == false:
```
Comment (Araq) [inline compiler/nimconf.nim]:
```suggestion
    if conf.projectFull.string != "" and not fileExists(conf.projectFull):
```

**Comment by sivchari** [comment]
```sh
  FAIL: telebot c
  Test "telebot" in category "nimble-packages"
  Failure: reBuildFailed
```
@Araq @timotheecour @xflywind 
Please tell me to pass the method to install above.

**Comment by timotheecour** [comment]
rebase against devel, i've disabled telebot in https://github.com/nim-lang/Nim/pull/17860 (but maybe telebot fixed the issue already, see https://github.com/ba0f3/telebot.nim/issues/59, so i've just restarted CI using approach described in contributing guide to see if it works)

File: compiler/nimconf.nim [hunk 2]
```diff
@@ -254,6 +254,10 @@ proc loadConfigs*(cfg: RelativeFile; cache: IdentCache; conf: ConfigRef; idgen:
     if isMain and optWasNimscript in conf.globalOptions:
       if conf.projectIsStdin: s = stdin.llStreamOpen
       elif conf.projectIsCmd: s = llStreamOpen(conf.cmdInput)
+    if conf.projectFull.string != "" and not fileExists(conf.projectFull):
+      if conf.cmd == cmdNimscript or conf.cmd == cmdUnknown:
+        echo "target file is not existed"
```
Comment (timotheecour) [inline compiler/nimconf.nim]: * not proper english (`doesn't exist` is better)
* should show relevant runtime context (`conf.projectFull`)
* don't use echo (stdout) for `errors`, use `localError`
* this PR makes `nim --eval:'echo 123'` worse by showing `target file is not existed`
* the bug isn't properly fixed:

```
nim foo $timn_D/tests/nim/all/t12217.nonexistant
target file is not existed
Error: invalid command: foo
```

instead should show:
```
nim foo /pathto/t12217.nonexistant
Error: nimscript file '/pathto/t12217.nonexistant' not found
```

> however, this is contradict to #17596. so, I think that we might have to talk about nimscript specification before this PR.

why? you just need to check whether the file exists

Comment (sivchari) [reply compiler/nimconf.nim]: thaks for reviewing.
I fixed codes.

Comment (sivchari) [reply compiler/nimconf.nim]: > why? you just need to check whether the file exists

I'm sorry, I misunderstood.
Please ignore it.

File: tests/misc/trunner.nim [hunk 3]
```diff
@@ -248,6 +248,17 @@ tests/newconfig/bar/mfoo.nims""".splitLines
     var expected = &"Hint: used config file '{filename}' [Conf]\31\n"
     doAssert outp.endsWith "123" & "\n" & expected
 
+  block: # exec NimScript
+    let filebase = testsDir / "newconfig/foo/"
```
Comment (timotheecour) [inline tests/misc/trunner.nim]:
```suggestion
    let dir = testsDir / "newconfig/foo"
```
and use `/` not `&` for joining paths

File: compiler/nimconf.nim [hunk 4]
```diff
@@ -254,6 +254,11 @@ proc loadConfigs*(cfg: RelativeFile; cache: IdentCache; conf: ConfigRef; idgen:
     if isMain and optWasNimscript in conf.globalOptions:
       if conf.projectIsStdin: s = stdin.llStreamOpen
       elif conf.projectIsCmd: s = llStreamOpen(conf.cmdInput)
+    if not fileExists(conf.projectFull):
+      # ex) pass --eval, existing Nimscript
+      if conf.cmd == cmdNimscript: discard
+      # ex) nim foo main.noextension
+      elif conf.cmd == cmdUnknown: rawMessage(conf, errGenerated, conf.projectFull.string() & " not found")
```
Comment (timotheecour) [inline compiler/nimconf.nim:261]:
```suggestion
      elif conf.cmd == cmdUnknown: rawMessage(conf, errGenerated, conf.projectFull.string & " not found")
```

Comment (timotheecour) [inline compiler/nimconf.nim]:
```suggestion
      if conf.cmd == cmdUnknown:
        # e.g. nim foo main.badext
        rawMessage(conf, errGenerated, conf.projectFull.string & " not found")
      else: # e.g. nim --eval:code
        discard
```

File: tests/misc/trunner.nim [hunk 5]
```diff
@@ -248,6 +248,17 @@ tests/newconfig/bar/mfoo.nims""".splitLines
...
+    let filenames = ["main.nims", "main.noextension", "main"]
+    let expectedArr = ["in task foo\n", "Error: " & filebase & "main.noextension not found\x1F\n", "Error: " & filebase & "main not found\x1F\n"]
+    let exitCodeArr = [0, 1, 1]
+    for i, filename in filenames:
+      var target = filebase & filename
```
Comment (timotheecour) [inline tests/misc/trunner.nim]: use let when possible

File: tests/misc/trunner.nim [hunk 6]
```diff
@@ -248,6 +248,17 @@ tests/newconfig/bar/mfoo.nims""".splitLines
...
+      var cmd = fmt"{nim} --hints:off foo {target}"
+      var expected = expectedArr[i]
```
Comment (timotheecour) [inline tests/misc/trunner.nim]: https://github.com/timotheecour/Nim/issues/534#issuecomment-780791150

**Comment by timotheecour** [comment]
this breaks `yaml` package, and you can reproduce this locally:

```
git clone https://github.com/flyx/NimYAML
cd NimYAML
nim build
```
`Error:  not found`

Only 1 important_packages package uses this though, and I don't like the fact that `nim build` was legal because it's too permissive and dependent on `cwd`, and this is made worse by the fact that any command is accepted depending on what's in `config.nims` (and can clash with future nim builtin commands but that's a different story).

The error message in this PR should be improved though, by detecting that no project was given and reporting this.

## before PR
```
nim build config.nims # work
nim build # works
nim build . # error
cd ..
nim build NimYAML/config.nims # works
nim build NimYAML # error
```

## after PR

```
nim build config.nims # work
nim build # error (as it should IMO, but the error is not informative)
nim build . # error (as it should IMO, but the error is misleading: `/pathto/NimYAML not found` even though that dir exists)
cd ..
nim build NimYAML/config.nims # works
nim build NimYAML # error
```

/cc @araq do you agree with turning `nim build` into an error ? (but improving the error message as mentioned above) ?

EDIT: alternatively, this PR can be modified to still allow the case where a project name was not specified and that case can be addressed in future work to unblock this PR (@sivchari  maybe let's do that?)

**Comment by Araq** [comment]
> /cc @Araq do you agree with turning nim build into an error ? (but improving the error message as mentioned above) ?

Fine with me.

**Comment by sivchari** [comment]
> EDIT: alternatively, this PR can be modified to still allow the case where a project name was not specified and that case can be addressed in future work to unblock this PR (@sivchari maybe let's do that?)

ok, If so, are the following actions correct?

```nim
nim build config.nims # work
nim build # error
nim build . # error
cd ..
nim build NimYAML/config.nims # works
nim build NimYAML # error
```

this can be achieved with this code.
```nim
if not fileExists(conf.projectFull):
      if conf.cmd == cmdUnknown: # e.g. nim foo main.badext
        if conf.projectName != "": # add this line
          rawMessage(conf, errGenerated, conf.projectFull.string & " not found")
      else: # e.g. nim --eval:code
        discard
```

> nim build . # error (as it should IMO, but the error is misleading: `/pathto/NimYAML not found` even though that dir exists)

I agree with you. 
~What kind of error message would be appropriate?~

EDIT: How about the following implementation?
cc. @Araq @timotheecour 
```nim
 if not fileExists(conf.projectFull):
      if conf.cmd == cmdUnknown: # e.g. nim foo main.badext
        if dirExists(conf.projectFull.string) or conf.projectName == "":
          rawMessage(conf, errGenerated, "failed to run. please check your command: nim" & conf.commandLine)
        rawMessage(conf, errGenerated, conf.projectFull.string & " not found")
      else: # e.g. nim --eval:code
        discard
```

```sh
nim build config.nims # work
nim build # Error: failed to run. please check your command: nim build
nim build . # Error: failed to run. please check your command: nim build .
cd ..
nim build NimYAML/config.nims # works
nim build NimYAML # Error: failed to run. please check your command: nim build NimYAML
```

**Comment by Clyybber** [comment]
> /cc @Araq do you agree with turning nim build into an error ? (but improving the error message as mentioned above) ?

This completely breaks the nimscript task system. Not every task must have an existing target file (might create one itself. think output file for documentation, etc.), or target any file for that matter (like the build task in NimYaml).

**Comment by timotheecour** [comment]
debatable that this should stay valid for reasons mentioned in https://github.com/nim-lang/Nim/pull/17847#issuecomment-831418012 (eg, normally nim commands don't depend on cwd, unlike with this case), but we can revisit this later, how about doing the more conservative change proposed earlier:

> alternatively, this PR can be modified to still allow the case where a project name was not specified and that case can be addressed in future work to unblock this PR

**Comment by sivchari** [comment]
@Clyybber How do you think ?

**Pull request closed without merging**